Kotlin Code Gen module

hzsweers opened the pull request:
This is an initial implementation of a code gen module. Originally from https://github.com/hzsweers/CatchUp/tree/master/tooling/moshkt

There's three main components:
* `moshi-kotlin-codegen` - the actually processor implementation
* `integration-test` - An integration test. This contains ignored KotlinJsonAdapter tests (Jesse mentioned wanting them to be able to pass the same tests is the long term goal, so this gives something to chip at) and some basic data classes tests.
* `moshi-kotlin-codegen-runtime` - A runtime artifact with the `@MoshiSerializable` annotation and its corresponding `MoshiSerializableJsonAdapterFactory`. The factory usage is optional, as one could write a factory generator if they wanted to.

Supported:
* Data classes
* `@Json` annotations
* Kotlin language features like nullability and default values (it generates Kotlin code via KotlinPoet, so it can actually leverage these features)
* If a companion object is specified on the source type, it generates an extension `jsonAdapter()` function onto it
* Generics
* Good chunk of Kotshi tests

Unimplemented:
- Support for more than just data classes
- JsonQualifier annotations

For data classes, it's been working swimmingly in CatchUp as well as @rharter's codebases. Code itself could probably use some cleaning up (there's plenty of TODOs left in), but its output seems to be working well so far.

CC @Takhion

Example:

```kotlin
@MoshiSerializable
data class Foo(
   @Json(name = "first_name") val firstName: String,
   @Json(name = "last_name") val lastName: String,
   val age: Int,
   val nationalities: List<String> = emptyList(),
   val weight: Float,
   val tattoos: Boolean = false,
   val race: String?,
   val hasChildren: Boolean = false,
   val favoriteFood: String? = null,
   val favoriteDrink: String? = "Water"
)
```

Generates

```kotlin
package com.squareup.moshi

import kotlin.Any
import kotlin.Boolean
import kotlin.Float
import kotlin.Int
import kotlin.String
import kotlin.collections.List

class Foo_JsonAdapter(moshi: Moshi) : JsonAdapter<Foo>() {
  private val string_Adapter: JsonAdapter<String> = moshi.adapter(String::class.java)

  private val int_Adapter: JsonAdapter<Int> = moshi.adapter(Int::class.java)

  private val list__string_Adapter: JsonAdapter<List<String>> = moshi.adapter<List<String>>(Types.newParameterizedType(List::class.java, String::class.java))

  private val float_Adapter: JsonAdapter<Float> = moshi.adapter(Float::class.java)

  private val boolean_Adapter: JsonAdapter<Boolean> = moshi.adapter(Boolean::class.java)

  override fun fromJson(reader: JsonReader): Foo? {
    if (reader.peek() == JsonReader.Token.NULL) {
      reader.nextNull<Any>()
    }
    lateinit var firstName: String
    lateinit var lastName: String
    var age = 0
    var nationalities: List<String>? = null
    var weight = 0.0f
    var tattoos: Boolean? = null
    var race: String? = null
    var hasChildren: Boolean? = null
    var favoriteFood: String? = null
    var favoriteDrink: String? = null
    reader.beginObject()
    while (reader.hasNext()) {
      when (reader.selectName(OPTIONS)) {
        0 -> firstName = string_Adapter.fromJson(reader)!!
        1 -> lastName = string_Adapter.fromJson(reader)!!
        2 -> age = int_Adapter.fromJson(reader)!!
        3 -> nationalities = list__string_Adapter.fromJson(reader)!!
        4 -> weight = float_Adapter.fromJson(reader)!!
        5 -> tattoos = boolean_Adapter.fromJson(reader)!!
        6 -> race = string_Adapter.fromJson(reader)
        7 -> hasChildren = boolean_Adapter.fromJson(reader)!!
        8 -> favoriteFood = string_Adapter.fromJson(reader)
        9 -> favoriteDrink = string_Adapter.fromJson(reader)
        -1 -> {
          // Unknown name, skip it
          reader.nextName()
          reader.skipValue()
        }
      }
    }
    reader.endObject()
    return Foo(firstName = firstName,
        lastName = lastName,
        age = age,
        weight = weight,
        race = race).let {
          it.copy(nationalities = nationalities ?: it.nationalities,
              tattoos = tattoos ?: it.tattoos,
              hasChildren = hasChildren ?: it.hasChildren,
              favoriteFood = favoriteFood ?: it.favoriteFood,
              favoriteDrink = favoriteDrink ?: it.favoriteDrink)
        }
  }

  override fun toJson(writer: JsonWriter, value: Foo?) {
    if (value == null) {
      writer.nullValue()
      return
    }
    writer.beginObject()
    writer.name("first_name")
    string_Adapter.toJson(writer, value.firstName)
    writer.name("last_name")
    string_Adapter.toJson(writer, value.lastName)
    writer.name("age")
    int_Adapter.toJson(writer, value.age)
    writer.name("nationalities")
    list__string_Adapter.toJson(writer, value.nationalities)
    writer.name("weight")
    float_Adapter.toJson(writer, value.weight)
    writer.name("tattoos")
    boolean_Adapter.toJson(writer, value.tattoos)
    if (value.race != null) {
      writer.name("race")
      string_Adapter.toJson(writer, value.race)
    }
    writer.name("hasChildren")
    boolean_Adapter.toJson(writer, value.hasChildren)
    if (value.favoriteFood != null) {
      writer.name("favoriteFood")
      string_Adapter.toJson(writer, value.favoriteFood)
    }
    if (value.favoriteDrink != null) {
      writer.name("favoriteDrink")
      string_Adapter.toJson(writer, value.favoriteDrink)
    }
    writer.endObject()
  }

  private companion object SelectOptions {
    private val OPTIONS: JsonReader.Options by lazy { JsonReader.Options.of("first_name", "last_name", "age", "nationalities", "weight", "tattoos", "race", "hasChildren", "favoriteFood", "favoriteDrink") }
  }
}
```

rharter commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,616 @@
+  private fun mavenGeneratedDir(adapterName: String): File {
+    // Hack since the maven plugin doesn't supply `kapt.kotlin.generated` option
+    // Bug filed at https://youtrack.jetbrains.com/issue/KT-22783
+    val file = filer.createSourceFile(adapterName).toUri().let(::File)
```
I was thinking about this more yesterday and realized that this only needs to be called once.  KotlinPoet takes care of the fqcn path, so this is going to return the same file every time.

The only time that might be different is for different source sets (test, main, etc), but I think those will get different processors (please confirm), so that shouldn't matter.

hzsweers replied:
Not sure I follow. Could you elaborate on the "called once" bit? Since we're only calling it "once" per file anyway 🤔

rharter replied:
We call it once per file, but we're getting the root source directory from it, so it only needs to be called once per source set. KotlinPoet takes care of the package folders.

FabianTerhorst commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,616 @@
+        optionsCN,
+        PRIVATE)
+        .delegate(
+            "lazy { %T.of(${optionsByIndex.map { it.value.key }
```
Are you sure the not very costly operation of creating some constants is worth an anonymous inner class for each MoshiSerializable data class adapter?

hzsweers replied:
I'm not! But creating options does quite a bit under the hood, and was hoping to gather feedback here

JakeWharton replied:
Classloading already makes this lazy. The explicitly lazy isn't needed.

hzsweers replied:
aa0534e

hzsweers commented:
Big ol' update above. 

* Generics support
* Lots of various bugfixes
* Introduction of Kotshi's test code

Questions that need deciding before moving forward:
1 - Should this support kotshi’s `@JsonDefaultValue` pattern? Or to what extent? I’m not totally familiar with it, but on the face of it I’m not sure it adds anything we can’t handle with the current approach
2 - Should this default to/support configurating primitive adapters (nextBoolean() vs .adapter(Boolean::class.java))?
3 - Do we want toString() implementations for adapters?
4 - Kotshi will use `@get:JvmName` over `@field:Json` if both are present. I don’t think this is idiomatic if we’re reading/generating kotlin code though, and should just keep it simple with `@Json`. Not sure if I’m missing context here though
5 - Currently we're using `lateinit` for expected non-nullable/non-defaultable fields. Wondering if we should just make them all nullable properties and then check the non-nullable/non-defaultable ones at the end to collect a list of all the missing ones. Would line up with something Kotshi appears to do as well
6 - Having to use a kotlinpoet snapshot to pick up the `Any?` bounds fix, so this can't merge till the next release anyway. Opportunistically picked up the `PropertySpec` literals support as well

hzsweers commented:
Also this is up for grabs if anyone has any ideas: d4d4926

Talked with @Takhion a bit, but haven't been able to arrive at a good solution yet to match the constructor from kotlin to java. Right now I fell back to just grabbing the first constructor.

Takhion commented:
>haven't been able to arrive at a good solution yet to match the constructor from kotlin to java

Before this PR gets merged I'll release a new version of `kotlin-metadata` that fixes all edge cases related to that, thanks to the excellent detective work from @hzsweers 🕵️  🙌

rharter commented:
For 1, I would say no. Unless I'm misunderstanding the `@JsonDefaultValue` duplicates Kotlin default values and is used because those aren't easily exposed in the Java code. IMHO they make the code very messy and verbose, losing some of the benefit of Kotlin.

hzsweers commented:
toString - 36522bc

default value handling - Agreed. Talked with Jesse offline too. 881d446

Remove custom names test, per talking with Jesse we'll just go with simple `@Json` usage - 8126bd1

Decided against configuration for primitive adapters - fb7e514

Just leaves this bit:

```
5 - Currently we're using lateinit for expected non-nullable/non-defaultable fields. Wondering if we should just make them all nullable properties and then check the non-nullable/non-defaultable ones at the end to collect a list of all the missing ones. Would line up with something Kotshi appears to do as well
```

Takhion commented:
About 5, my personal opinion is: absolutely 👍

hzsweers commented:
Good enough for me! I'll work on that. Was maybe a bit too excited about using local `lateinit` from kotlin 1.2 :)

JakeWharton commented:
They're implemented the same way so it doesn't really matter.

hzsweers commented:
Implemented tracking missing props in 3fd6614

This makes all nullable handling for local properties the same, and removes defaults for primitives in the process. It simplifies the handling a lot, and leans on kotlin language features to take care of null handling (null checking and then throwing the lazily evaluated list of missing properties).

One minor change from what kotshi does - this reports the serialized name in the missing properties, not the property name. We could look at supporting this though if we want. It also throws JsonDataException rather than NPEs.

hzsweers commented:
Implemented JsonQualifier support in e8427be! I made `Types.createJsonQualifierImplementation()` public to support it, let me know what you think of that.

Resulting code looks like this:

```
private val string_Adapter_for_WrappedInArray_WrappedInObject_Adapter: JsonAdapter<String> =
            moshi.adapter<String>(String::class.java, setOf(Types.createJsonQualifierImplementation(WrappedInArray::class.java), Types.createJsonQualifierImplementation(WrappedInObject::class.java))).nullSafe()
```

Pretty happy with the result - no extra methods, passes the kotshi tests. It does rely on proxies, but I'm not sure there's a way around that right now since kotlin classes don't support subclassing and the only alternative I can think of is to reflectively go yank them out of the constructor params at runtime 😷 .

Not sure what the CI failure is, maybe travis goofed? Works for me locally 🤔

hzsweers commented:
Rebased and also switched to using kapt+autoservice for the processor service declaration. I think this is ready for a serious review now, as I think it's feature-complete for data classes in terms of generated code.

Missing bits:
- Decision on how to handle multiple constructors. Right now it assumes one and picks the first. Supporting a specific one will require an updated kotlin-metadata with improved signature matching
- Will need to wait till the next kotlinpoet release since it's using a snapshot right now

swankjesse commented:
First off, this is awesome. Thank you. Sorry for the slow response; I‘ve had a few things going on that’s prevented me from spending as much time with code as I’d prefer.

I’m going to start by reviewing the generated code. I think that’s one of the most interesting.

```
class Simple_JsonAdapter(moshi: Moshi) : JsonAdapter<Simple>() {
```
Accepting a `moshi` instance in the constructor is clever. Dig.

Not sure about the underscore in the generated class name. Is `SimpleJsonAdapter` okay?

```
    private val string_Adapter_Adapter: JsonAdapter<String> =
            moshi.adapter(String::class.java).nullSafe()
```
Why `string_Adapter_Adapter` and not `string_Adapter`? Or just `stringAdapter` ?

Why `.nullSafe()` ? Would it be better to omit `.nullSafe()` unless the target type is nullable?

```
        if (reader.peek() == JsonReader.Token.NULL) {
            reader.nextNull<Any>()
        }
```
This shouldn’t be necessary. The caller of this adapter can use `.nullSafe()` on this if they want to permit nulls.

The `when` clause paired with `Options` is really great. Love that.

```
        val missingArguments: () -> JsonDataException = {
                NAMES.filterIndexed { index, _ ->
                    when (index) {
                        0 -> prop == null
                        else -> false
                    }
                }.let { JsonDataException("The following required properties were missing: ${it.joinToString()}") }
                }
        if (prop == null) {
            throw missingArguments()
        }
```
This is nice behavior, but it’s so much code. How do you feel about this?
```
        if (prop == null) {
            throw JsonDataException("required property 'prop' missing at ${reader.path}")
        }
```
You won’t get a full list of missing arguments, but I think that’s okay. 

```
    override fun toJson(writer: JsonWriter, value: Simple?) {
        if (value == null) {
            writer.nullValue()
            return
        }
```
We should assume `value` is non-null. Callers should use `.nullSafe()` on this if they may have nulls.

```

    private companion object SelectOptions {
        private val NAMES: Array<String> = arrayOf("prop")

        private val OPTIONS: JsonReader.Options = JsonReader.Options.of(*NAMES)
    }
```
Nice use of options. But I think if you can live with not including the full list of missing parameters you can omit the `NAMES` declaration.

Also surprising to have a named companion object. I’d prefer just this:
```
    companion object {
        private val OPTIONS: JsonReader.Options = JsonReader.Options.of("prop")
    }
```

So TL;DR: this is how I’d have the generated code look. Mostly the same as what you already have, but generating less code.

```
class SimpleJsonAdapter(moshi: Moshi) : JsonAdapter<Simple>() {
    private val stringAdapter: JsonAdapter<String> =
            moshi.adapter(String::class.java)

    override fun toString(): String = "GeneratedJsonAdapter(Simple)"

    override fun fromJson(reader: JsonReader): Simple {
        var prop: String? = null

        reader.beginObject()
        while (reader.hasNext()) {
            when (reader.selectName(OPTIONS)) {
                0 -> prop = stringAdapter.fromJson(reader)
                -1 -> {
                    // Unknown name, skip it
                    reader.nextName()
                    reader.skipValue()
                }
            }
        }
        reader.endObject()

        if (prop == null) {
            throw JsonDataException("required property 'prop' missing at ${reader.path}")
        }

        return Simple(prop = prop)
    }

    override fun toJson(writer: JsonWriter, value: Simple?) {
        writer.beginObject()
        writer.name("prop")
        stringAdapter.toJson(writer, value!!.prop)
        writer.endObject()
    }

    companion object {
        private val OPTIONS: JsonReader.Options = JsonReader.Options.of("prop")
    }
}
```

swankjesse commented:
One more idea on the generated code: users will typically have a very small number of instances of each `Adapter` class: typically 1. I wonder if it’s simpler to omit the `companion object` altogether and declare the `options` as a normal instance-level val.

swankjesse commented:
From `WithCompanionFunction_JsonAdapter`:
```
        if (value.v != null) {
            writer.name("v")
            string_nullable_Adapter_Adapter.toJson(writer, value.v)
        }
```
I think the `if()` check is unnecessary. We can just emit the null `"v"` unconditionally and let Moshi omit it if `serializeNulls()` is enabled.

swankjesse commented:
From `MultipleJsonQualifiers_JsonAdapter`:
```
    private val string_Adapter_for_WrappedInArray_WrappedInObject_Adapter: JsonAdapter<String> =
            moshi.adapter<String>(String::class.java, setOf(Types.createJsonQualifierImplementation(WrappedInArray::class.java), Types.createJsonQualifierImplementation(WrappedInObject::class.java))).nullSafe()
```
I’m reluctant to expose the weird `Types.createJsonQualifierImplementation` method. Instead should we just add an overload of `Moshi.adapter()` that takes a varargs of annotation types? That might be a lot easier to call.

swankjesse commented:
One more idea for the null checks: the generated code _could_ do this and omit the `if (prop == null)` checks altogether:

```
        return Simple(
                prop = prop  ?: throw JsonDataException("required property 'prop' missing at ${reader.path}")
        )
```

swankjesse commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,116 @@
+annotation class MoshiSerializable
+
+class MoshiSerializableFactory : JsonAdapter.Factory {
+  private val adapters = Collections.synchronizedMap(
```
I don’t think this is necessary. Moshi will cache these itself.

hzsweers replied:
c95a2b9

swankjesse commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,116 @@
+      return adapterCtor
+    }
+    val clsName = cls.name.replace("$", "_")
+    if (clsName.startsWith("android.")
```
I don’t think this check should be necessary. These classes won’t have `@MoshiSerializable` !

hzsweers replied:
c69c477

swankjesse commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  companion object {
+    private var instance: WeakReference<MoshiSerializableFactory>? = null
```
Ick. Why? This is Moshi competing with Dagger/Guice, etc.

hzsweers replied:
Ha, easy to change. Do you prefer a Singleton instance or a static factory?

swankjesse replied:
Ummm could it just be a constructor?

hzsweers replied:
I suppooooose

hzsweers replied:
4a91af8

swankjesse commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,116 @@
+      }
+
+    } catch (e: ClassNotFoundException) {
+      adapterCtor = findConstructorForClass(cls.superclass)
```
I find this surprising. Presumably this doesn’t work for deserialization; it would return the wrong type

hzsweers replied:
I think this might be a vestige of the buterknife code this is adapted from. Agreed it shouldn't be necessary

hzsweers replied:
b9b2527

swankjesse commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,116 @@
+
+@Retention(RUNTIME)
+@Target(CLASS)
+annotation class MoshiSerializable
```
Should we fold the small runtime component into Moshi itself? Maybe even install the `MoshiSerializableFactory` by default? It’d need to be rewritten in .java, but that’s not a big deal.

swankjesse replied:
... I’m imagining like `@JsonObject` or something to sit alongside `@Json` as part of the core API.

hzsweers replied:
Can definitely do that, but worried people might use it and not realize they need to wire in the compiler (especially since it would only fail at runtime). What do you think?

swankjesse replied:
That's a problem regardless. We can fail at runtime with a nice message if they mess up.

hzsweers commented:
> I’m reluctant to expose the weird Types.createJsonQualifierImplementation method. Instead should we just add an overload of Moshi.adapter() that takes a varargs of annotation types? That might be a lot easier to call.

Agreed! Much better - 7461d68

> I think the if() check is unnecessary. We can just emit the null "v" unconditionally and let Moshi omit it if serializeNulls() is enabled.

dea9c0b

>  I wonder if it’s simpler to omit the companion object altogether and declare the options as a normal instance-level val.

Makes sense! 1f08c70

> One more idea for the null checks: the generated code could do this and omit the if (prop == null) checks altogether:

Chatted more offline, result is differentiation between absent and null in 428917d

> We should assume value is non-null. Callers should use .nullSafe() on this if they may have nulls.

Removed from reads in 90436da

> You won’t get a full list of missing arguments, but I think that’s okay.

Cleaned up and inlined the exceptions thrown into the construction in 8eb5bb4

Also simplified adapter names in d6e8a84 and 4132192

JakeWharton commented on kotlin-codegen/compiler/pom.xml:
```diff
@@ -0,0 +1,134 @@
+    <repository>
+      <id>jcenter</id>
+      <url>https://jcenter.bintray.com/</url>
+    </repository>
```
This will prevent us from releasing to Maven central.

hzsweers replied:
I hadn't thought of that. What do you suggest? I don't know if @Takhion has strong feelings on publishing to mavencentral (this is for kotlin-metadata)

Takhion replied:
Quite the opposite! I've already received other requests to publish to Maven central, and the next version will be there for sure (soon) :+1:

hzsweers replied:
perf

JakeWharton commented on kotlin-codegen/compiler/pom.xml:
```diff
@@ -0,0 +1,134 @@
+    <dependency>
+      <groupId>org.jetbrains.kotlin</groupId>
+      <artifactId>kotlin-stdlib</artifactId>
+    </dependency>
```
Move up so compile and test artifacts are grouped

hzsweers replied:
dc8647f

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+    return if (roundEnv.getElementsAnnotatedWith(annotationElement)
+            .asSequence()
+            .mapNotNull { processElement(it) }
+            .any { !it.generateAndWrite() }) true else true
```
This is functional abuse. Break out into a `forEach` and simply have `return true`

hzsweers replied:
leftover copy pasta from the kotlin-metadata sample I started from :P.  a33d46c

JakeWharton commented on moshi/src/main/java/com/squareup/moshi/Moshi.java:
```diff
@@ -70,6 +71,20 @@
+  @CheckReturnValue
+  public <T> JsonAdapter<T> adapter(Type type, Class<? extends Annotation>... annotationTypes) {
+    if (annotationTypes.length == 1) {
+      // Save the iterator
```
There is no iterator for arrays

hzsweers replied:
derp 917ae9b

JakeWharton commented on moshi/src/main/java/com/squareup/moshi/Types.java:
```diff
@@ -172,7 +172,8 @@ static Type canonicalize(Type type) {
   @SuppressWarnings("unchecked")
-  static <T extends Annotation> T createJsonQualifierImplementation(final Class<T> annotationType) {
+  static <T extends Annotation> T createJsonQualifierImplementation(
+      final Class<T> annotationType) {
```
Revert?

hzsweers replied:
62a9f1b

JakeWharton commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,86 @@
+
+class MoshiSerializableFactory : JsonAdapter.Factory {
+
+  override fun create(type: Type, annotations: MutableSet<out Annotation>,
```
`Set`

hzsweers replied:
ef4cdd6

JakeWharton commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,86 @@
+      return if (constructor.parameterTypes.size == 1) {
+        constructor.newInstance(moshi)
+      } else {
+        if (type is ParameterizedType) {
```
collapse to `else if` to reduce nesting

hzsweers replied:
This let me turn it into a `when` - 9445169

JakeWharton commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/kotshi/test/TestPrimitiveDefaultValues.kt:
```diff
@@ -0,0 +1,80 @@
+            adapter.toJson(this, actual)
+          }
+        }
+        .readUtf8())
```
This is incredibly hard to read. Can we just use locals instead of `apply` and `run`?

hzsweers replied:
Agreed, this was copied code from the kotshi tests. a987697

JakeWharton commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/kotshi/test/TestClassWithJavaKeyword.kt:
```diff
@@ -0,0 +1,47 @@
+                indent = "  "
+              }, ClassWithJavaKeyword(true, 4711, 1337))
+        }
+        .readUtf8()
```
Hard to read. Use locals

hzsweers replied:
12857e1

JakeWharton commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/kotshi/test/TestAdapterGeneration.kt:
```diff
@@ -0,0 +1,194 @@
+                        adapter.toJson(this, actual)
+                    }
+                }
+                .readUtf8())
```
Here too

hzsweers replied:
449667b

JakeWharton commented on kotlin-codegen/integration-test/pom.xml:
```diff
@@ -0,0 +1,153 @@
+    <dependency>
+      <groupId>org.jetbrains.kotlin</groupId>
+      <artifactId>kotlin-reflect</artifactId>
+    </dependency>
```
Move these two up above test deps

hzsweers replied:
Part of dc8647f

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+                    }
+                  }
+                  Projection.STAR -> WildcardTypeName.subtypeOf(ANY)
+                  Projection.INV -> TODO("INV projection is unsupported")
```
Is this an actual TODO or should it be `error(..)`?

hzsweers replied:
I actually don't know. I forgot about that and meant to ask what you want to do with this case

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+        .addFunction(FunSpec.builder("fromJson")
+            .addModifiers(OVERRIDE)
+            .addParameter(reader)
+            .returns(originalTypeName.asNullable())
```
No reason to put `asNullable()` since we tell the caller to call `nullSafe()` if they want `null` support.

hzsweers replied:
960f85e

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+                addStatement("%N.name(%S)",
+                    writer,
+                    prop.serializedName)
+                addStatement("%N.toJson(%N, %N${if (index == 0) "!!" else ""}.%L)",
```
No need to emit `!!`. Just mark the parameter as non-null.

hzsweers replied:
@swankjesse and I talked about this a bit. That might work in gradle, but I don't see how to do that in a maven project. It doesn't see the JSR305 bit, so it fails to compile if you make the parameter nonnull

JakeWharton replied:
The only way that it would fail to compile is if it did see a `@Nullable` annotation directly on the parameter.

JakeWharton replied:
There is a `@Nullable` annotation on the parameter.

I would just emit a manual null check as the first statement then rather than complicate the implementation with this logic.

hzsweers replied:
Fair enough. Went with a more explicit nullcheck but let me know if you'd rather it just be a `value!!` at the beginning. 5ddf528

JakeWharton replied:
Nope. Like what you got.

hzsweers replied:
:+1:

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+      // If it's an Array type, we shortcut this to return Types.arrayOf()
+      if (rawType == ARRAY) {
+        return CodeBlock.of("%T.arrayOf(%L)",
+            Types::class.asTypeName(),
```
General heads-up: KotlinPoet understands KClass and so you don't need to box instances when emitting to a `%T`

hzsweers replied:
Neat, I wasn't sure. Did some cleanup in 0390cc5

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+                  }
+            }
+            .beginControlFlow("-1 ->")
+            .addCode("// Unknown name, skip it\n")
```
nit: Period at the end of comment. [Comment should be sentences][1].

 [1]: https://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

hzsweers replied:
0702fc7

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+    is TypeVariableName -> {
+      CodeBlock.of("%N[%L]", typesArray, genericTypeNames.indexOfFirst { it == this })
+    }
+  // Shouldn't happen
```
Remove. The exception already says this.

hzsweers replied:
0702fc7

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+    }
+    is WildcardTypeName -> "wildcard__" + (lowerBounds + upperBounds).simplifiedNames()
+    is TypeVariableName -> name.decapitalize() + if (bounds.isEmpty()) "" else "__" + bounds.simplifiedNames()
+  // Shouldn't happen
```
Throw?

hzsweers replied:
e090521

JakeWharton commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,759 @@
+  return when (this) {
+    is ClassName -> this
+    is ParameterizedTypeName -> rawType
+    else -> throw UnsupportedOperationException("Cannot get raw type from $this")
```
IllegalArgumentException?

hzsweers replied:
I always wondered if that still made sense in extension functions 🤔 . I guess it does!

hzsweers replied:
1fec18a

NightlyNexus commented:
Can you propose the Moshi.adapter API addition in a separate PR?

hzsweers commented:
I can, but any particular reason to? It would presumably just link back to this as an example use case

Egorand commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,760 @@
+
+    val originalTypeName = originalElement.asType().asTypeName()
+    val moshiName = "moshi".allocate()
+    val moshiParam = ParameterSpec.builder(moshiName, Moshi::class.asClassName()).build()
```
`asClassName()` looks redundant, you should be able to simply pass `Moshi::class`

hzsweers replied:
9a34bb0

Egorand commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,760 @@
+                  }
+            }
+            .beginControlFlow("-1 ->")
+            .addCode("// Unknown name, skip it.\n")
```
`addComment()` can add `\\` and `\n` for you

hzsweers replied:
Ahh good catch! 7525f89

ansman commented on kotlin-codegen/runtime/src/main/java/com/squareup/moshi/MoshiSerializable.kt:
```diff
@@ -0,0 +1,81 @@
+@Target(CLASS)
+annotation class MoshiSerializable
+
+class MoshiSerializableFactory : JsonAdapter.Factory {
```
Why not make this an `object`?

hzsweers replied:
I asked higher up about whether or not to make this class a singleton (lazy or otherwise), and we went with just making this instantiable

swankjesse replied:
It’s consistent with what we did for reflection.
KotlinJsonAdapterFactory

NightlyNexus commented on moshi/src/main/java/com/squareup/moshi/Moshi.java:
```diff
@@ -70,6 +71,19 @@
+    for (Class<? extends Annotation> annotationType : annotationTypes) {
+      annotations.add(Types.createJsonQualifierImplementation(annotationType));
+    }
+    return adapter(type, annotations);
```
`Collections.unmodifiableMap`

hzsweers replied:
d877e88

NightlyNexus commented on moshi/src/main/java/com/squareup/moshi/Moshi.java:
```diff
@@ -70,6 +71,19 @@
+  @CheckReturnValue
+  public <T> JsonAdapter<T> adapter(Type type, Class<? extends Annotation>... annotationTypes) {
+    if (annotationTypes.length == 1) {
+      return adapter(type,
```
maybe `return adapter(type, annotationTypes[0])`
It's interesting that we know `annotationTypes.length` is never 0 because of the existing method signature.

hzsweers replied:
That's clever, and saves a bit of code duplication - 5ed2a58

NightlyNexus commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,760 @@
+                  qualifiers.isEmpty() -> "" to emptyArray()
+                  qualifiers.size == 1 -> {
+                    ", %${standardArgsSize}T::class.java" to arrayOf(
+                        qualifiers.first().annotationType.asTypeName())
```
this looks like it will run into https://github.com/ansman/kotshi/issues/60 (same below)

hzsweers replied:
Interesting. Definitely interested in seeing Moshi support that case

NightlyNexus replied:
Moshi does support it. Both the KotlinJsonAdapter and the ClassJsonAdapter hand off the annotation instances to `adapter(Type, Set...)`.

hzsweers replied:
that's via reflectively pulling off the annotation instances, whereas the idea here was to avoid that since we know them upfront

NightlyNexus replied:
This PR does `adapter(Type, Class<?>`). This will create a new annotation instance that loses the declared annotation elements.

hzsweers replied:
I see what you mean, but I'm confused on what you mean by `Moshi does support it` above? One way I could envision that case working here is by reflectively pulling them off the parameters at runtime, as kotlin annotations don't allow for something like `AutoAnnotation` since they're `final` :/ . 

The other, albeit possibly crazy way, is to generate proxy invocation handlers that include the statically defined data (kind of our own inline kotlin-ified autoannotation with proxies)

NightlyNexus commented on moshi/src/main/java/com/squareup/moshi/Moshi.java:
```diff
@@ -70,6 +71,19 @@
   }
 
+  @CheckReturnValue
+  public <T> JsonAdapter<T> adapter(Type type, Class<? extends Annotation>... annotationTypes) {
```
in light of my comment on needing the JsonQualifier elements, i'm hoping this isn't needed, at least for this PR.

hzsweers replied:
As in you'd prefer to see this API removed? Or that you're hoping needing jsonqualifier elements support isn't needed for this PR?

NightlyNexus replied:
I'm thinking that this API won't be needed because the usage of it won't be the intended behavior (honoring the annotation elements).

hzsweers replied:
I don't quite follow. Why doesn't this?

hzsweers commented:
Yeah I didn't intend for this PR to become this big (wanted to keep it to the processor and iterate on integration tests), but with the offline requests to add both the kotlinjsonadapter tests and kotshi tests, I don't see how it could have stayed small

NightlyNexus commented:
in toJson, there's a check for non-null types:
```kotlin
if (value == null) {
  throw NullPointerException("value was null! Wrap in .nullSafe() to write nullable values.")
}
```
should there be a symmetric check in fromJson (peek on the reader for null)?

In `DataClassesTest_DefaultValuesJsonAdapter.kt`, I got this:
```kotlin
return DataClassesTest.DefaultValues(foo = foo ?: throw JsonDataException("Required property 'foo' missing at ${reader.path}"))
                .let {
                  it.copy(bar = bar ?: it.bar,
                      nullableBar = if (nullableBarSet) nullableBar else it.nullableBar,
                      bazList = bazList ?: it.bazList)
                }
```
Do we need the `copy`?

In `DataClassesTest_DefaultValuesJsonAdapter`, the `list__stringAdapter` property could lose the duplicate underscore.

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,760 @@
+ * If you don't want this though, you can use the runtime [MoshiSerializable] factory implementation.
+ */
+@AutoService(Processor::class)
+class MoshiKotlinCodeGenProcessor : KotlinAbstractProcessor(), KotlinMetadataUtils {
```
Mostly adopted from the example in the repo, but I'm not exactly in favor of just copying in the utilities we need (implicit fork, etc). I don't really see an issue with this if we're ok with otherwise using `AbstractProcessor()` as a base class

hzsweers replied:
Yeah I agree, we can clean it up in a followup. Will talk with @Takhion

hzsweers commented:
Trying to unpack the larger comment above and answering inline, let me know if I've messed up the groupings.

> in toJson, there's a check for non-null types:

There was, but not anymore after dea9c0b. Also - the code snippet seems to be from `fromJson()`? I don't think it ever threw exceptions on null values in `toJson()`

> should there be a symmetric check in fromJson (peek on the reader for null)?

This went through a few iterations in the PR, but the most recent one we landed on after @swankjesse's past pass was to read everything into nullable local variables, then throw eagerly in the event of null/absent values for require properties in 8eb5bb4

> Do we need the copy?

Yes. Worked on this with @rharter and @swankjesse a fair bit before the PR, it's basically a clever way to "read" the default values from the object since we can't really read them at compile time and inline them.

> the list__stringAdapter property could lose the duplicate underscore.

I agree! Wasn't sure how to differentiate type params in the simplified names, suggestions welcome :). General trend from the other naming cleanups was to move as close as possible to camel casing and general readability.  

> it'd still be neat to generate a JsonAdapter.Factory instead of using this class. or was there an explicit reason not to?

(I oddly can't respond to this one inline on github 🤔)
This pattern scales pretty horribly in multi module apps. The idea behind the approach here is to keep the library agnostic with the optional runtime factory for basic drop-in use. Someone else could still write their own factory generator if they want too!

Also - assuming you meant to comment on the factory and not the annotation? We definitely need the annotation in some form as a sentinel to the annotation processor.

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/KotlinCodeGenTest.kt:
```diff
@@ -0,0 +1,760 @@
+    A, B
+  }
+
+  @Ignore @Test fun interfacesNotSupported() {
```
nice

swankjesse replied:
... does this need `@MoshiSerializable` on the interface?

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/KotlinCodeGenTest.kt:
```diff
@@ -0,0 +1,760 @@
+    }
+  }
+
+  abstract class AbstractClass(val a: Int)
```
same

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/KotlinCodeGenTest.kt:
```diff
@@ -0,0 +1,760 @@
+  inner class InnerClass(val a: Int)
+
+  @Ignore @Test fun localClassesNotSupported() {
+    class LocalClass(val a: Int)
```
@MoshiSerializable ?

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/KotlinCodeGenTest.kt:
```diff
@@ -0,0 +1,760 @@
+    }
+  }
+
+  @Ignore @Test fun objectDeclarationsNotSupported() {
```
oooh neat

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/kotshi/ClassWithWeirdNames.kt:
```diff
@@ -0,0 +1,15 @@
+import com.squareup.moshi.MoshiSerializable
+
+@MoshiSerializable
+data class ClassWithWeirdNames(
```
love it

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/kotshi/DefaultValues.kt:
```diff
@@ -0,0 +1,50 @@
+data class WithCompanionProperty(val v: String?)
+
+@MoshiSerializable
+data class WithStaticProperty(val v: String?)
```
huh?

swankjesse replied:
what’s the static property?

hzsweers replied:
Might have been a vestige of the kotshi default values API

swankjesse commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/kotshi/GenericClass.kt:
```diff
@@ -0,0 +1,7 @@
+package com.squareup.moshi.kotshi
```
I’m reluctant to merge this unless it’s a contribution from the original author. CLA stuff.

swankjesse replied:
@ansman ?

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+    val classData = metadata.data
+    val (nameResolver, classProto) = classData
+
+    fun ProtoBuf.Type.extractFullName() = extractFullName(classData)
```
copy-pasta from the sample processor I originally started from. Could be an extension val or just a local lambda too, which were you thinking?

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+        .enclosedElements
+        .mapNotNull { it.takeIf { it.kind == ElementKind.CONSTRUCTOR }?.let { it as ExecutableElement } }
+        .first()
+    // TODO Temporary until jvm method signature matching is better
```
I think @Takhion said he has a WIP fix for this

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+        hasCompanionObject = hasCompanionObject,
+        visibility = classProto.visibility!!,
+        genericTypeNames = genericTypeNames,
+        elementUtils = elementUtils)
```
I think it was a leftover from prototyping, can definitely look to try to remove this (which I think you did in your followup PR)

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+    }
+    is WildcardTypeName -> "wildcard__" + (lowerBounds + upperBounds).simplifiedNames()
+    is TypeVariableName -> name.decapitalize() + if (bounds.isEmpty()) "" else "__" + bounds.simplifiedNames()
+    else -> throw IllegalArgumentException("Unrecognized type! $this")
```
That works too. I almost kind of wish that TypeName was a `sealed` class in kotlinpoet

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+      rawType.simpleName().decapitalize() + if (typeArguments.isEmpty()) "" else "__" + typeArguments.simplifiedNames()
+    }
+    is WildcardTypeName -> "wildcard__" + (lowerBounds + upperBounds).simplifiedNames()
+    is TypeVariableName -> name.decapitalize() + if (bounds.isEmpty()) "" else "__" + bounds.simplifiedNames()
```
nice

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+  }.let { if (nullable) "${it}_nullable" else it }
+}
+
+private fun ClassName.isClass(elementUtils: Elements): Boolean {
```
I think this is actually supposed to be `isInterface()` 😥

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+
+private fun TypeName.makeType(
+    elementUtils: Elements,
+    typesArray: ParameterSpec,
```
this is the `Type[]` parameter for generics. Admittedly confusing name

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+    val unaliasedName: TypeName,
+    val jsonQualifiers: Set<AnnotationMirror>) {
+
+  val isRequired = !nullable && !hasDefault
```
it could be. Felt like one of those cases where it could go either way and I opted for precomputing as a calculated property

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+        .associate { prop ->
+          val typeName = prop.unaliasedName
+          val qualifierNames = prop.jsonQualifiers.joinToString("") {
+            "at${it.annotationType.asElement().simpleName.toString().capitalize()}"
```
can do

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+          }
+        }
+        .apply {
+          // TODO make this configurable. Right now it just matches the source model
```
Probably fine to start out. My main thought was cases where the model is public but the generated adapter is internal

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+  }
+}
+
+private fun ProtoBuf.TypeParameter.asTypeName(
```
I'd be down

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+ */
+private fun ProtoBuf.Type.asTypeName(
+    nameResolver: NameResolver,
+    getTypeParameter: (index: Int) -> ProtoBuf.TypeParameter,
```
this is admittedly running with what was in the kotlin-metadata example and not questioning it 🙃

hzsweers commented on kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt:
```diff
@@ -0,0 +1,764 @@
+
+  val argumentList = when {
+    hasAbbreviatedType() -> abbreviatedType.argumentList
+    else -> argumentList
```
reads it off of the typealias list if this is an alias

hzsweers commented on kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/KotlinCodeGenTest.kt:
```diff
@@ -0,0 +1,760 @@
+import kotlin.annotation.AnnotationRetention.RUNTIME
+
+class KotlinCodeGenTest {
+  @Ignore @Test fun constructorParameters() {
```
Can add one. This is from when the initial goal was to try to make the `KotlinJsonAdapter` tests pass with this. That's also why most of the below tests are ignored

hzsweers commented:
Think I covered the questions you had. Once #462 is merged, I can make a followup with some cleanups from above and also fro #461
